### ISSUE TITLE
Sentiment RNN predict() function predicting wrong results with negative reviews

### ISSUE THREAD

**Issue by saadtazi** [udacity/deep-learning-v2-pytorch#5]
Hi, in the RNN sentiment analysis solution (https://github.com/udacity/deep-learning-v2-pytorch/blob/master/sentiment-rnn/Sentiment_RNN_Solution.ipynb), when I use the predict() function provided in the notebook with negative reviews (even the ones in the training/valid/test data), I don't get the expected result most of the time (accuracy: 1/4). Here is the code I added in a cell at the end of the notebook:

```
# from https://github.com/udacity/deep-learning-v2-pytorch/blob/f456d6167d8af205154cd27f0bcfe5eb813e2503/sentiment-rnn/Sentiment_RNN_Solution.ipynb
# at the end of the notebook...

rev1= 'this film is one giant pant load . paul schrader is utterly lost in his own bad screenplay . and his directing is about as comatose as it can be without his actually having been sleepwalking during the process .  br    br   the worst though is woody harrelson  whom i ordinarily like when he  s properly cast . he plays  the walker   a homosexual man in d . c . who plays social companion to the bored wives of the washington elite . he couldn  t have been more one dimensional if he had been cut out of a magazine and bounced around in front of the camera on a popsicle stick . his  southern accent  is that  off the rack  version that decrescendos from the beginning to the end of every line he delivers  as though the heat and humidity of the south is still draining him of every ounce of energy he has . it is monotonous . but  his is not the worst accent in the movie . his  boyfriend   played by moritz bleibtreau  attempts to affect some kind of a mid east accent that is so clumsy he can barely deliver the bad lines written for him . he is incapable of rolling his r  s in spite of the fact that in real life he is german  and speaks several languages  one of them being italian  that  s kind of a good reason to cast someone else don  t ya think   br    br   from the story  to the screenplay  to the directing  to the camera work  to the performances by the leads  this movie is bad from beginning to end . the only tolerable moments in this film came from three supporting actresses lily tomlin  lauren bacall  and kristin scott thomas . only these three managed to make it through this movie with their dignity in tact . in fact  all three are excellent  in spite of being trapped in a really bad film . ufortunately  no one could ever be good enough to redeem this endless series of flaws . if you like these three actresses  watch them in something else . this movie is not worth your time .'
rev2 = "when i was little my parents took me along to the theater to see interiors . it was one of many movies i watched with my parents  but this was the only one we walked out of . since then i had never seen interiors until just recently  and i could have lived out the rest of my life without it . what a pretentious  ponderous  and painfully boring piece of    s wine and cheese tripe . woody allen is one of my favorite directors but interiors is by far the worst piece of crap of his career . in the unmistakable style of ingmar berman  allen gives us a dark  angular  muted  insight in to the lives of a family wrought by the psychological damage caused by divorce  estrangement  career  love  non  love  halitosis  whatever . the film  intentionally  has no comic relief  no music  and is drenched in shadowy pathos . this film style can be best defined as expressionist in nature  using an improvisational method of dialogue to illicit a  more pronounced depth of meaning and truth  . but woody allen is no ingmar bergman . the film is painfully slow and dull . but beyond that  i simply had no connection with or sympathy for any of the characters . instead i felt only contempt for this parade of shuffling  whining  nicotine stained  martyrs in a perpetual quest for identity . amid a backdrop of cosmopolitan affluence and baked brie intelligentsia the story looms like a fart in the room . everyone speaks in affected platitudes and elevated language between cigarettes . everyone is  lost  and  struggling   desperate to find direction or understanding or whatever and it just goes on and on to the point where you just want to slap all of them . it  s never about resolution  it  s only about interminable introspective babble . it is nothing more than a psychological drama taken to an extreme beyond the audience  s ability to connect . woody allen chose to make characters so immersed in themselves we feel left out . and for that reason i found this movie painfully self indulgent and spiritually draining . i see what he was going for but his insistence on promoting his message through prozac prose and distorted film techniques jettisons it past the point of relevance . i highly recommend this one if you  re feeling a little too happy and need something to remind you of death . otherwise  let  s just pretend this film never happened ."
rev3 = "this film lacked something i couldn  t put my finger on at first charisma on the part of the leading actress . this inevitably translated to lack of chemistry when she shared the screen with her leading man . even the romantic scenes came across as being merely the actors at play . it could very well have been the director who miscalculated what he needed from the actors . i just don  t know .  br    br   but could it have been the screenplay  just exactly who was the chef in love with  he seemed more enamored of his culinary skills and restaurant  and ultimately of himself and his youthful exploits  than of anybody or anything else . he never convinced me he was in love with the princess .  br    br   i was disappointed in this movie . but  don  t forget it was nominated for an oscar  so judge for yourself ."
rev4 = "sorry everyone    i know this is supposed to be an  art  film   but wow  they should have handed out guns at the screening so people could blow their brains out and not watch . although the scene design and photographic direction was excellent  this story is too painful to watch . the absence of a sound track was brutal . the loooonnnnng shots were too long . how long can you watch two people just sitting there and talking  especially when the dialogue is two people complaining . i really had a hard time just getting through this film . the performances were excellent  but how much of that dark  sombre  uninspired  stuff can you take  the only thing i liked was maureen stapleton and her red dress and dancing scene . otherwise this was a ripoff of bergman . and i  m no fan f his either . i think anyone who says they enjoyed     hours of this is   well  lying . "

# seq_length = 200
print("rev1", predict(net, rev1, seq_length))
print("rev2", predict(net, rev2, seq_length))
print("rev3", predict(net, rev3, seq_length))
print("rev4", predict(net, rev4, seq_length))

# output:
# rev1 1
# rev2 0
# rev3 1
# rev4 1
# mostly ones...
```
(also available here: https://gist.github.com/saadtazi/7e641645fc90b3d547165ce4e2f10800)

Any idea why?

**Comment by cezannec** [comment]
I get the same behavior, my guess is that the trained model actually isn't that great! It's only trained for 2 epochs and may even be too large of a model for simple dataset (and at risk of overfitting based on certain keywords). I'd suggest decreasing the hidden sizes of the layers, and training for a longer time, perhaps with a smaller learning rate, to get a model that is better at inference. Thank you for pointing this out; I will add this as a note in the classroom!

**Comment by jaketlussier** [comment]
Hi! I ran into the same thing as @saadtazi, and it seems to me that the model from the notebook is memorizing the alternating class structure. If you shuffle the data, for example, all bets are off! Testing accuracy degrades from perfect to random. Or if you instead shift the testing data by one (so that you start with a negative instead of a positive), testing performance will be 0%!

Example snippets you can try adding in:

`
# Before padding...
import random
zipped = list(zip(reviews_ints, encoded_labels))
random.shuffle(zipped)
review_ints, encoded_labels = zip(*zipped)
encoded_labels = np.array(encoded_labels)
`

`
# After setting batch_size=49 in testing cell...
inputs = inputs[1:, :]
labels = labels[1:]
`

Haven't had time to debug myself, but it seems like init_hidden() should perhaps be called for each batch, or there may even need to memory clearing within batches.

**Comment by jaketlussier** [comment]
Hi! I ran into the same thing as @saadtazi, and it seems to me that the model from the notebook is memorizing the alternating class structure. If you shuffle the data, for example, all bets are off! Testing accuracy degrades from perfect to random. Or if you instead shift the testing data by one (so that you start with a negative instead of a positive), testing performance will be 0%!

Example snippets you can try adding in:

```python
import random
zipped = list(zip(reviews_ints, encoded_labels))
random.shuffle(zipped)
review_ints, encoded_labels = zip(*zipped)
encoded_labels = np.array(encoded_labels)
```

```python
# After setting batch_size=49 in testing cell...
inputs = inputs[1:, :]
labels = labels[1:]
```

Haven't had time to debug myself, but it seems like init_hidden() should perhaps be called for each batch, or there may even need to memory clearing within batches.

**Comment by hpkhanh1610** [comment]
In my opinion, there may be some issue related to the **forward()** function. I just opened a new issue related to that (https://github.com/udacity/deep-learning-v2-pytorch/issues/9). Please check the following carefully, some students are also confusing in this section:

# reshape to (hidden_dim, batch_size, n_classes)
sig_out = sig_out.view(-1, batch_size, self.output_size)
sig_out = sig_out[0] # get last batch of labels

**Comment by cezannec** [comment]
Ah, I see, I never shuffled the data before training! I am pretty sure that the hidden state is behaving correctly, but it could definitely be a problem with the forward function as well. 

I'll shuffle the data and work on a new solution today. Thanks for this helpful information!

**Comment by cezannec** [comment]
Shuffled the data and updated how I shaped my data in the forward function. The solution is now updated and the inference behavior is also much better as a result! I tested on both positive and negative reviews thanks to your comments :)
